src/test/java/com/google/impactdashboard/server/ProjectInformationRetrieverTest.java

```diff
@@ -0,0 +1,113 @@
+
+  public static final long PROJECT_NUMBER_1 = 12345678L;
+  public static final long PROJECT_NUMBER_2 = 98765432543L;
+
+  public static final double PROJECT_AVERAGEIAMBINDINGS_1 = 162.3;
+  public static final double PROJECT_AVERAGEIAMBINDINGS_2 = 2000.87;
+
+  public static final String RECOMMENDATION_DESCRIPTION = "This is a recommendation";
+
+  public static final Map<Long, Integer> PROJECT_IAM_DATA = new HashMap<>();
+  public static final Map<Long, Recommendation> PROJECT_RECOMMENDATION_DATA = new HashMap<>();
+
+  private DataReadManager readManager;
+  private ProjectInformationRetriever informationRetriever;
+
+  @Before
+  public void setup() {
+    readManager = mock(DataReadManager.class);
+    informationRetriever = ProjectInformationRetriever.create(readManager);
+  }
+
+  @Test
+  public void emptyProjectInformationTest() {
+    // There is not information so there should be an empty list returned
+    when(readManager.listProjects()).thenReturn(Collections.emptyList());
+    List<Project> actual = informationRetriever.listProjectInformation();
+    List<Project> expected = new ArrayList<>();
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void nonEmptyProjectInformationTest() {
+    // There will be two projects returned from the DataBase
+    // Expected output will be two projects
+    when(readManager.listProjects()).thenReturn(Arrays.asList(
+        ProjectIdentification.create(PROJECT_NAME_1, PROJECT_ID_1, PROJECT_NUMBER_1),
+        ProjectIdentification.create(PROJECT_NAME_2, PROJECT_ID_2, PROJECT_NUMBER_2)
+    ));
+    when(readManager.getAverageIAMBindingsInPastYear(PROJECT_ID_1)).thenReturn(PROJECT_AVERAGEIAMBINDINGS_1);
+    when(readManager.getAverageIAMBindingsInPastYear(PROJECT_ID_2)).thenReturn(PROJECT_AVERAGEIAMBINDINGS_2);
+
+    List<Project> expected = Arrays.asList(
+        Project.create(PROJECT_NAME_1, PROJECT_ID_1, PROJECT_NUMBER_1,
+            ProjectMetaData.create(PROJECT_AVERAGEIAMBINDINGS_1)),
+        Project.create(PROJECT_NAME_2, PROJECT_ID_2, PROJECT_NUMBER_2,
+            ProjectMetaData.create(PROJECT_AVERAGEIAMBINDINGS_2)));
+    List<Project> actual = informationRetriever.listProjectInformation();
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void projectWithNoData() {
+    // Project exists but there is no data for the project
+    when(readManager.getMapOfDatesToIAMBindings(PROJECT_ID_1)).thenReturn(Collections.emptyMap());
+    when(readManager.getMapOfDatesToRecommendationTaken(PROJECT_ID_1))
+        .thenReturn(Collections.emptyMap());
+
+    ProjectGraphData expected = ProjectGraphData.create(PROJECT_ID_1, Collections.emptyMap(),
+        Collections.emptyMap());
+    ProjectGraphData actual = informationRetriever.getProjectData(PROJECT_ID_1);
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void projectWithData() {
+    // Project exists and data is present
+    PROJECT_IAM_DATA.put(123L, 2000);
```

Comment:
It's better to not modify global test data inside a particular test, because it means that other tests written in the future would have to be written with that understanding as well. 

Instead, a local variable is much better in this scenario.

Reply:
That definitely seems like a better way to do this. I guess I was just thinking everything that I'm using should be static and final but that doesn't need to be true.